pkg/streamingpromql/operators/binary_operation.go

```diff
@@ -507,52 +568,245 @@ func (b *BinaryOperation) mergeOneSide(data []types.InstantVectorSeriesData, sou
+	// We don't expect to have many series here, so something like a loser tree is likely unnecessary.
+	for {
+		if remainingSeriesWithHistograms == 1 {
+			// Only one series left, just copy remaining points.
+			for _, d := range data {
+				if len(d.Histograms) > 0 {
+					output = append(output, d.Histograms...)
+					return output, nil
+				}
+			}
+		}
+
+		nextT := int64(math.MaxInt64)
+		sourceSeriesIndexInData := -1
+
+		for seriesIndexInData, d := range data {
+			if len(d.Histograms) == 0 {
+				continue
+			}
+
+			nextPointInSeries := d.Histograms[0]
+			if nextPointInSeries.T == nextT {
+				// Another series has a point with the same timestamp. We have a conflict.
+				firstConflictingSeriesLabels := ds.sourceSeriesMetadata[ds.sourceSeriesIndices[sourceSeriesIndexInData]].Labels
+				secondConflictingSeriesLabels := ds.sourceSeriesMetadata[ds.sourceSeriesIndices[seriesIndexInData]].Labels
+				groupLabels := b.labelsFunc()(firstConflictingSeriesLabels)
+
+				return nil, fmt.Errorf("found duplicate series for the match group %s on the %s side of the operation at timestamp %s: %s and %s", groupLabels, side, timestamp.Time(nextT).Format(time.RFC3339Nano), firstConflictingSeriesLabels, secondConflictingSeriesLabels)
+			}
+
+			if d.Histograms[0].T < nextT {
+				nextT = d.Histograms[0].T
+				sourceSeriesIndexInData = seriesIndexInData
+			}
+		}
+
+		output = append(output, data[sourceSeriesIndexInData].Histograms[0])
+		data[sourceSeriesIndexInData].Histograms = data[sourceSeriesIndexInData].Histograms[1:]
+
+		if len(data[sourceSeriesIndexInData].Histograms) == 0 {
+			remainingSeriesWithHistograms--
+		}
 	}
 }
 
+func (b *BinaryOperation) computeResult(left types.InstantVectorSeriesData, right types.InstantVectorSeriesData) (types.InstantVectorSeriesData, error) {
+	var fPoints []promql.FPoint
+	var hPoints []promql.HPoint
+
+	// For one-to-one matching for arithmetic operators, we'll never produce more points than the smaller input side.
+	// Because histograms and histograms can be multiplied together, we use the sum of both the histogram and histogram points.
+	// We also don't know if the output will be exclusively histograms or histograms, so we'll use the same size slice for both.
+	// We only assign the slices once we see the associated point type so it shouldn't be common that we allocate both.
+	//
+	// FIXME: this is not safe to do for one-to-many, many-to-one or many-to-many matching, as we may need the input series for later output series.
+	canReturnLeftFPointSlice, canReturnLeftHPointSlice, canReturnRightFPointSlice, canReturnRightHPointSlice := true, true, true, true
+	leftPoints := len(left.Floats) + len(left.Histograms)
+	rightPoints := len(right.Floats) + len(right.Histograms)
+	maxPoints := max(leftPoints, rightPoints)
+
+	prepareFSlice := func() error {
+		if maxPoints <= len(left.Floats) && len(left.Floats) < len(right.Floats) {
```

Comment:
Just realised: we should use the capacity of the slices here, not the lengths, as the slices might only have a handful of elements but have capacity for plenty more.